src/libstd/keyword_docs.rs

```diff
@@ -1578,12 +1578,157 @@ mod type_keyword {}
+/// and `unsafe impl`, but also `unsafe fn` -- see below). They are not mutually
+/// exclusive, as can be seen in `unsafe fn`.
+///
+/// Unsafe Rust adds some capabilities to Safe Rust:
+///
+/// - Dereference [raw pointers]
+/// - Call `unsafe` functions (including foreign functions, some compiler
+/// intrinsics and the raw allocator)
+/// - Implement `unsafe` [`trait`]s
+/// - Mutate [`static`]s (including [`extern`]al ones)
+/// - Access fields of [`union`]s
+///
+/// # Why do we need a keyword for this ?
+///
+/// **No matter what, Safe Rust can't cause Undefined Behavior**. This is
+/// referenced to as [soundness]: a well-typed program actually has the desired
+/// properties. In Safe Rust, the language and the compiler do this mostly by
+/// themselves whereas in Unsafe Rust the programmer must help. The [Nomicon]
+/// has a more detailed explanation on the subject.
+///
+/// # The different meanings of `unsafe`
+///
+/// Depending on the place where it is used, the `unsafe` keyword meaning can
+/// change. See the following [discussion on Rust Internals] for more in-depth
+/// explanations.
+///
+/// Summarizing the main points:
+///
+/// - `unsafe fn`: calling this function means abiding by a contract the
+/// compiler cannot enforce.
+/// - `unsafe trait Trait`: implementing this [`trait`] means abiding by a
+/// contract the compiler cannot enforce.
+/// - `unsafe {}`: the contract necessary to call the operations inside the
+/// block has been checked and is guaranteed to be respected.
+/// - `unsafe impl`: the contract necessary to implement the trait has been
+/// checked and is guaranteed to be respected.
+///
+/// As of now (Rust 1.44), `unsafe fn` also acts like an `unsafe {}` block
+/// around the code inside the function. This means it is not just a signal to
+/// the caller, but also promises that the preconditions for the operations
+/// inside the function are upheld.
+///
+/// Mixing these two meanings can be confusing and [proposal]s exist to use
+/// `unsafe {}` blocks inside such functions when making `unsafe` operations.
+///
+/// # Dereferencing raw pointers
+///
+/// ```rust
+/// let a = 3;
+/// let b = &a as *const _;
+/// // SAFETY: a has not been dropped and references are always aligned,
+/// // so b is a valid address.
+/// let _ = unsafe { *b };
+/// ```
+///
+/// # Marking elements as unsafe
+///
+/// `unsafe` can be used on functions. Note that functions declared in
+/// [`extern`] blocks are implicitly marked as `unsafe`. Methods can also be
+/// declared as `unsafe`:
+///
+/// ```rust
+/// # #![allow(dead_code)]
+/// unsafe fn unsafe_fn() {}
+///
+/// extern "C" {
+///     fn unsafe_extern_fn();
+/// }
+///
+/// trait TraitWithUnsafeMethod {
+///     unsafe fn unsafe_method();
+/// }
 ///
+/// struct S;
+///
+/// impl S {
+///     unsafe fn unsafe_method_on_struct() {}
+/// }
+/// ```
+///
+/// Traits can also be declared as `unsafe`:
```

Comment:
Does `unsafe trait` functions automatically become unsafe?

Reply:
[This playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=41ff0e9586df80004509c044978e76e7) confirms it (it compiles and run successfully) confirms they do not. Is it surprising enough to warrant a line though ?

Reply:
Not sure about that but I think we should talk about it. I don't know about it but it does surprised me.

Reply:
No they do not and should not become `unsafe fn`. `unsafe trait` and `unsafe fn` are orthogonal dimensions.

`unsafe trait` means "Implementations promise to uphold the extra guarantees described in the trait docs". But the methods in that trait can still be safe to call! There might just be extra guarantees. Like:
```
/// Implementations guarantee that `num` always returns an even integer.
unsafe trait ToNum {
  fn num(self) -> i32;
}
```
Here, `num` is always safe to call, but additionally callers may rely on it returning something even. I would make no sense to make it an `unsafe fn`.

Reply:
@poliorcetics We could demonstrate this through showing what's inside the `unsafe trait` (we probably still need to keep an empty unsafe trait to demonstrate marking).

```rust
unsafe trait Extra {
    // unsafe fn needs to be specified explicitly like normal functions
    unsafe fn a();

    // unsafe trait have safe functions except being marked as unsafe fn
    fn a();
}
```

Reply:
I mixed your two ideas to produce a `## Methods inside unsafe trait` subsection so I'll resolve this and further discussion can start from this new subsection.